src/app/forms/components/defect/defect.component.ts

```diff
@@ -1,85 +1,191 @@
+    private dfs: DynamicFormService,
+    private pipe: DefaultNullOrEmpty,
+    private router: Router,
+    private testResultsStore: Store<TestResultsState>,
+    private resultService: ResultOfTestService,
+    private errorService: GlobalErrorService
+  ) {
+    this.isEditing = this.activatedRoute.snapshot.data['isEditing'];
+  }
+
+  ngOnInit(): void {
+    const defectIndex = this.testResultsStore.pipe(select(selectRouteParam('defectIndex')));
+    const defectRef = this.testResultsStore.pipe(select(selectRouteParam('ref')));
+
+    this.testResultsStore
+      .select(toEditOrNotToEdit)
+      .pipe(
+        withLatestFrom(defectIndex, defectRef),
+        takeUntil(this.onDestroy$),
+        filter(([testResult]) => !!testResult)
+      )
+      .subscribe(([testResult, defectIndex, defectRef]) => {
+        this.defects = testResult!.testTypes[0].defects;
+        this.vehicleType = testResult!.vehicleType;
+        this._defectsForm = (this.dfs.createForm(DefectsTpl, testResult) as CustomFormGroup).get(['testTypes', '0', 'defects']) as CustomFormArray;
+        if (defectIndex) {
+          this.index = Number(defectIndex!);
+          this.form = this._defectsForm.controls[this.index] as CustomFormGroup;
+          this.defect = this.defects![this.index];
+        } else if (defectRef) {
+          this.defectsStore
+            .select(selectByDeficiencyRef(defectRef, this.vehicleType))
+            .pipe(take(1))
+            .subscribe(([defect, item, deficiency]) => {
+              this.initializeDefect(defect as Defect, item as Item, deficiency as Deficiency);
+            });
         }
       });
+
+    this.defectsStore.select(selectByImNumber(this.defect!.imNumber!, this.vehicleType)).subscribe(defectsTaxonomy => {
+      this.initializeInfoDictionary(defectsTaxonomy);
+    });
+  }
+
+  ngOnDestroy(): void {
+    this.onDestroy$.next(true);
+    this.onDestroy$.complete();
+  }
+
+  get isDangerous(): boolean {
+    return this.defect!.deficiencyCategory === 'dangerous';
+  }
+
+  get isAdvisory(): Boolean {
+    return this.defect!.deficiencyCategory === 'advisory';
+  }
+
+  handleSubmit() {
+    const errors: GlobalError[] = [];
+    DynamicFormService.updateValidity(this.form, errors);
+
+    if (errors.length > 0) {
+      this.errorService.setErrors(errors);
+    }
+
+    if (this.form.invalid) {
+      return;
+    }
+
+    if (this.index || this.index === 0) {
+      this.testResultsStore.dispatch(updateDefect({ defect: this.form.getCleanValue(this.form) as TestResultDefect, index: this.index }));
+    } else {
+      this.testResultsStore.dispatch(createDefect({ defect: this.form.getCleanValue(this.form) as TestResultDefect }));
     }
+    this.resultService.updateResultOfTest();
+    this.navigateBack();
   }
 
-  @Output() removeDefect = new EventEmitter<number>();
+  handleRemove() {
+    this.testResultsStore.dispatch(removeDefect({ index: this.index }));
+    this.resultService.updateResultOfTest();
```

Comment:
move the `updateResultOfTest` call to the `navigateBack` function since we want to call it when we call that function and remove and submit anyways